### DISCUSSION TITLE
Multiple event loops (macOS)

### DISCUSSION THREAD

**Discussion by savvn001** [hbldh/bleak#1191] [Q&A] [answered]
Hi, 

I think I'm trying to do the same thing as in this example: https://github.com/hbldh/bleak/pull/1034#issue-1386945656 ie. I want to have multiple threads running, that each have their own instance of a `BleakClient`, allowing simultaneous connections to different devices, but each in their own thread.

From what I can gather, this was an issue, but fixed for BlueZ backend, but what about macOS (coreBluetooth)? 

When I do the same I get the `got Future <Future pending> attached to a different loop` error upon calling `await client.connect()`.

**Comment by savvn001** [comment]
Update: 

I think I found a slightly-janky-but-works workaround:

Inside the thread where I want the client to run, I just dig into the centralManagerDelegate and manually set a new event loop for this thread. I think it was set before, to a different event loop, from the other thread that was previously scanning.

```
event_loop = asyncio.new_event_loop()
client = BleakClient(device, disconnected_callback=self.__disconnect_cb)

# Workaround, set event loop inside CentralManagerDelegate
client._backend._central_manager_delegate.event_loop = event_loop
```

**Comment by dlech** [comment]
That example was just to make unit tests run cleanly. There should never be a need to have multiple run loops. Why can't all connections just use a single run loop?

**Comment by savvn001** [comment, marked as answer]
Actually yes, this is wrong thinking on my part.
The design paradigm for bleak is to have multiple connections all under 1 event loop in 1 thread right, (rather than each device has it's own instance of bleak)

Part of the reason I went down this thinking is because bleak is working in the backend of a Plotly Dash app, where each button press for example, starts a callback in a new thread. So the scanner was started in 1 thread, then a new connection in another thread, so 2 different event loops, causing this issue.

I think the correct design pattern, would be all the dash threads send queue messages to a single bleak thread to do everything, in 1 event loop.

**Reply by dlech** [reply]
Yes, that would be the way to go.
